fiftyone/brain/internal/core/leaky_splits.py

```diff
@@ -0,0 +1,497 @@
+
+
+def _to_views(samples, split_views=None, split_field=None, split_tags=None):
+    """Helper function so that we can always work with views"""
+
+    arithmetic_true = lambda x: int(x is not None)
+    num_given = (
+        arithmetic_true(split_views)
+        + arithmetic_true(split_field)
+        + arithmetic_true(split_tags)
+    )
+
+    if num_given == 0:
+        raise ValueError(f"One of the split arguments must be given.")
+    if num_given > 1:
+        raise ValueError(f"Only one of the split arguments must be given.")
+
+    if split_views:
+        return split_views
+
+    if split_field:
+        return _field_to_views(samples, split_field)
+
+    if split_tags:
+        return _tags_to_views(samples, split_tags)
+
+
+def _field_to_views(samples, field):
+    field_values = samples.distinct(field)
+
+    if len(field_values) < 2:
+        raise ValueError(
+            f"Field {field} has less than 2 distinct values,"
+            f"can't be used to create splits"
+        )
+
+    views = []
+    for val in field_values:
+        view = samples.match(F(field) == val)
+        views.append(view)
+
+    return views
+
+
+def _tags_to_views(samples, tags):
+    if len(tags) < 2:
+        raise ValueError("Must provide at least two tags.")
+
+    views = []
+    for tag in tags:
+        view = samples.match_tags([tag])
+        views.append(view)
+    return views
+
+
+###
+
+### SKL BACKEND
+class LeakySplitsSKLConfig(
+    skl_sim.SklearnSimilarityConfig, LeakySplitsConfigInterface
+):
+    """Configuration for Leaky Splits with the SKLearn backend
+
+    Args:
+        split_views (None): list of views corresponding to different splits
+        split_field (None): field name that contains the split that the sample belongs to
+        split_tags (None): list of tags that correspond to different splits
+
+        For the rest of the arguments, see :class:`SklearnSimilarityConfig`
+    """
+
+    def __init__(
+        self,
+        split_views=None,
+        split_field=None,
+        split_tags=None,
+        embeddings_field=None,
+        model=None,
+        model_kwargs=None,
+        patches_field=None,
+        metric="cosine",
+        **kwargs,
+    ):
+        LeakySplitsConfigInterface.__init__(
+            self, split_views, split_field, split_tags
+        )
+        skl_sim.SklearnSimilarityConfig.__init__(
+            self,
+            embeddings_field=embeddings_field,
+            model=model,
+            model_kwargs=model_kwargs,
+            patches_field=patches_field,
+            metric=metric,
+            **kwargs,
+        )
+
+    @property
+    def method(self):
+        return "Neural"
+
+
+class LeakySplitsSKL(skl_sim.SklearnSimilarity):
+    def initialize(self, samples, brain_key):
+        return LeakySplitsSKLIndex(
+            samples, self.config, brain_key, backend=self
+        )
+
+
+class LeakySplitsSKLIndex(
+    skl_sim.SklearnSimilarityIndex, LeakySplitIndexInterface
```

Comment:
I'm not sure that subclassing from `SklearnSimilarityIndex` makes sense here. It has the side effect of causing a full copy of all of the embeddings to be stored on the dataset when `save_run_results()` is called in the main method.

I would instead argue that "leaky splits" is an algorithm that *may use* embeddings, but it shouldn't be directly responsible for storing a copy of the embeddings. Those should only be stored exactly once by `compute_similarity()`.

Downstream methods like this should be able to use pre-computed embeddings. Or, if no embeddings are provided, then this method can compute its own embeddings, but they won't actually be stored permanently. 

For context, this is how `compute_uniqueness()` and `compute_representativeness()` and `compute_visualization()` all work.

Reply:
@brimoor I mostly agree with this, but I think the default behavior of not storing the embeddings when the algorithm is run is potentially very frustrating for users, who need to recompute millions of embeddings.

Reply:
Yeah I get it. On the other hand, another source of annoyance would be having multiple copies of a potentially large number of embeddings duplicately stored on your dataset. And if you don't explicitly specify a model to use, you don't even know what the embeddings are as the default model used is not currently made known to the user.

The current mental model is basically:
- Provide a simple syntax `compute_uniqueness(dataset)` with minimal parameters that gives you the insight without concerning you with how it happens necessarily
- Provide advanced syntaxes `compute_uniqueness(dataset, **kwargs)` that allow for reusing embeddings, customizing the model used, etc.

Reply:
It's fine to use an `SklearnSimilarityIndex` instance internally (which could be a pre-existing one that users pass in), I just don't think you should subclass from it, because that has side effects like causing the embeddings to be stored permanently in the database under the `brain_key`.

Reply:
This was a huge sticking point for me when designing this, I'm glad it came up. I avoided holding an instance of the similarity index (despite it being more general because it allows the user to define what similarity index they want to use as long as it supports the duplicates mixin) because having a brain method running in the background without the user's knowledge seemed strange (wouldn't the instance still show up if the user listed all brain methods?).

Reply:
After meeting with Jacob:
1. Don't subclass from similarity
2. Have two arguments, brain_key and similarity_brain_key
3. if neither is passed, nothing is registered
4. if similarity_brain_key is passed and exists, load that run, otherwise compute and register
5. if similarity_brain_key is not passed, create but don't register similarity index
6. brain_key says if/where to save leaky splits run

Reply:
@brimoor can you please let us know what you think about the above refactor?